doc/tutorial/machine_learning_map/pyparsing.py

```diff
@@ -939,109 +1141,128 @@ def __getnewargs__(self):
+    on parsing strings containing C{<TAB>}s, and suggested methods to maintain a
+    consistent view of the parsed string, the parse location, and line and column
+    positions within the parsed string.
+    """
+    return strg.count("\n", 0, loc) + 1
+
+
+def line(loc, strg):
+    """Returns the line of text containing loc within a string, counting newlines as line separators."""
     lastCR = strg.rfind("\n", 0, loc)
     nextCR = strg.find("\n", loc)
     if nextCR >= 0:
-        return strg[lastCR+1:nextCR]
+        return strg[lastCR + 1 : nextCR]
     else:
-        return strg[lastCR+1:]
+        return strg[lastCR + 1 :]
+
 
-def _defaultStartDebugAction( instring, loc, expr ):
-    print (("Match " + _ustr(expr) + " at loc " + _ustr(loc) + "(%d,%d)" % ( lineno(loc,instring), col(loc,instring) )))
+def _defaultStartDebugAction(instring, loc, expr):
+    print(
+        (
+            "Match "
+            + _ustr(expr)
+            + " at loc "
+            + _ustr(loc)
+            + "(%d,%d)" % (lineno(loc, instring), col(loc, instring))
+        )
+    )
 
-def _defaultSuccessDebugAction( instring, startloc, endloc, expr, toks ):
-    print ("Matched " + _ustr(expr) + " -> " + str(toks.asList()))
 
-def _defaultExceptionDebugAction( instring, loc, expr, exc ):
-    print ("Exception raised:" + _ustr(exc))
+def _defaultSuccessDebugAction(instring, startloc, endloc, expr, toks):
+    print("Matched " + _ustr(expr) + " -> " + str(toks.asList()))
+
+
+def _defaultExceptionDebugAction(instring, loc, expr, exc):
+    print("Exception raised:" + _ustr(exc))
+
 
 def nullDebugAction(*args):
     """'Do-nothing' debug action, to suppress debugging output during parsing."""
     pass
 
+
 # Only works on Python 3.x - nonlocal is toxic to Python 2 installs
-#~ 'decorator to trim function calls to match the arity of the target'
-#~ def _trim_arity(func, maxargs=3):
-    #~ if func in singleArgBuiltins:
-        #~ return lambda s,l,t: func(t)
-    #~ limit = 0
-    #~ foundArity = False
-    #~ def wrapper(*args):
-        #~ nonlocal limit,foundArity
-        #~ while 1:
-            #~ try:
-                #~ ret = func(*args[limit:])
-                #~ foundArity = True
-                #~ return ret
-            #~ except TypeError:
-                #~ if limit == maxargs or foundArity:
-                    #~ raise
-                #~ limit += 1
-                #~ continue
-    #~ return wrapper
+# ~ 'decorator to trim function calls to match the arity of the target'
+# ~ def _trim_arity(func, maxargs=3):
+# ~ if func in singleArgBuiltins:
+# ~ return lambda s,l,t: func(t)
+# ~ limit = 0
+# ~ foundArity = False
+# ~ def wrapper(*args):
+# ~ nonlocal limit,foundArity
+# ~ while 1:
+# ~ try:
+# ~ ret = func(*args[limit:])
+# ~ foundArity = True
+# ~ return ret
+# ~ except TypeError:
+# ~ if limit == maxargs or foundArity:
+# ~ raise
+# ~ limit += 1
+# ~ continue
+# ~ return wrapper
```

Comment:
This can be removed.

Reply:
Not sure about this one:

The function that is commented out is for Python 3:
https://github.com/scikit-learn/scikit-learn/blob/d3429c138a1eab162f3627e14c6ac26f49d59a16/doc/tutorial/machine_learning_map/pyparsing.py#L994-L996

The current function is for Python 2-3:
https://github.com/scikit-learn/scikit-learn/blob/d3429c138a1eab162f3627e14c6ac26f49d59a16/doc/tutorial/machine_learning_map/pyparsing.py#L1015-L1017

Shouldn't we discard the current Python 2-3 code and enable the commented out Python 3 code?

Reply:
I feel this should be handled in a different PR.

Reply:
That's a good question. Even though Python 2 is not supported anymore, a conservative choice would be to keep the current one if it works for Python 3. But the commented one used for Python 3 is way smaller and using it this would ease maintenance.

In any case, I am not sure if tutorial are referenced nor used anymore.

Cc: @glemaitre for a more authoritative point of view.